fix: Keep sync component count in collection card [FC-0062]

ChrisChV opened the pull request:
<!--

Note: Please refer to the Support Development Guidelines on the wiki page to consider backporting to active releases:
https://openedx.atlassian.net/wiki/spaces/COMM/pages/4248436737/Support+Guidelines+for+active+releases

Please give your pull request a short but descriptive title.
Use conventional commits to separate and summarize commits logically:
https://open-edx-proposals.readthedocs.io/en/latest/oep-0051-bp-conventional-commits.html

Use this template as a guide. Omit sections that don't apply.
You may link to information rather than copy it, but only if the link is publicly readable.
If the linked information must be private (because it contains secrets), clearly label the link as private.

-->

## Description

Fix: 
- https://github.com/openedx/modular-learning/issues/234
- https://github.com/openedx/edx-platform/issues/35776

Fixed component counter synchronization in these cases:

1. When deleting a component inside a collection.
2. With the library published, when adding a new component in a collection and reverting library changes.
3. With the library published, when deleting a component inside a collection and reverting library changes.

Also adds a `published > num_counts` field in collections in the search index.

## Supporting information

Related github issues: 
- https://github.com/openedx/modular-learning/issues/234
- https://github.com/openedx/edx-platform/issues/35776
Internal link: https://tasks.opencraft.com/browse/FAL-3921
Depends on: https://github.com/openedx/openedx-learning/pull/257

## Testing instructions

- Run `make requirements` on lms and cms shell.

1. When deleting a component inside a collection

- Go the library home of a library
- Go to the collection list page.
- Create a component in a collection or use a collection with components.
- Delete a component on the collection home page.
- Return to the collection list page and verify the component count in the collection card

2. With the library published, when adding a new component in a collection and reverting library changes.

- Go the library home of a **published** library
- Go to the collection list page.
- Create a component in a collection on the collection home page.
- Return to the collection list page and verify the component count in the collection card.
- Discarg library changes
- Return to the collection list page and verify the component count in the collection card.

3. With the library published, when deleting a component inside a collection and reverting library changes.

- Go the library home of a **published** library
- Go to the collection list page.
- Create a component in a collection (and publish) or use a collection with components published.
- Delete a component on the collection home page
- Return to the collection list page and verify the component count in the collection card.
- Discarg library changes
- Return to the collection list page and verify the component count in the collection card.

To test the new  `published > num_counts` field, you can follow the testing instructions in: https://github.com/openedx/frontend-app-authoring/pull/1481

openedx-webhooks commented:
Thanks for the pull request, @ChrisChV!

### What's next?

*Please work through the following steps to get your changes ready for engineering review:*

#### :radio_button: Get product approval

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.

#### :radio_button: Provide context

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads





#### :radio_button: Get a green build

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.



#### :radio_button: Let us know that your PR is ready for review:

### Who will review my changes?


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.  Tag them in a comment and let them know that your changes are ready for review.


### Where can I find more information?

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)

### When can I expect my changes to be merged?

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*

<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

DanielVZ96 commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1327,6 +1327,11 @@ def revert_changes(library_key):
     authoring_api.reset_drafts_to_published(learning_package.id)
 
+    # If a component is created within a collection and then returned to the published
+    # version, that component must be removed from the collection.
```
```suggestion
    # version, that discarded component must be removed from the collection.
```

DanielVZ96 reviewed (approved):
:+1: LGTM just a small suggestion to make the comment easier to understand.

- [x] I tested this: (describe what you tested)
- [x] I read through the code
- [x] I checked for accessibility issues
- [ ] Includes documentation
- [x] I made sure any change in configuration variables is reflected in the corresponding client's `configuration-secure` repository.

bradenmacdonald commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -485,6 +486,19 @@ def searchable_doc_for_collection(
+            if draft:
+                draf_num_children += 1
+            if published:
+                published_num_children += 1
```
@ChrisChV CC @ormsbee

If a collection contains 200 items, will this result in 400 queries? I think we need a way to just get this as

```suggestion
        draft_num_children = collection.entities.filter(has_draft).count()
        published_num_children =collection.entities.filter(has_published).count()
```

Is that `filter` possible?

ormsbee replied:
That filter is definitely possible, but we might want to split it out a bit to give the publishing API control over translating from PublishableEntities -> Drafts... Something like

```python
draft_num_children = get_drafts(collection.entities.all()).count()
```

Where `get_drafts` is implemented in the `publishing` app. So ostensibly anything can throw in its qset of PublishableEntities and get Drafts without `collections` needing to know how that mapping really works.

ChrisChV replied:
Updated https://github.com/openedx/edx-platform/pull/35734/commits/b6f2d98a02bdf5e2c3be7062685b0f56a3b2bfee, using https://github.com/openedx/openedx-learning/pull/257

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -684,7 +686,12 @@ def _get_library_component_tags_count(library_key) -> dict:
+    library_key,
+    text_search=None,
+    block_types=None,
+    in_collections=False,
```
Personally I don't think this `in_collections` parameter should be part of this public API interface. It's a very specific thing that I can't really imagine other uses for. See below.

ChrisChV replied:
Updated: https://github.com/openedx/edx-platform/pull/35734/commits/5e771cebf28e066e462fa39e6dac79d713de79f4

bradenmacdonald commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1320,6 +1349,35 @@ def revert_changes(library_key):
+    # Use case: When a component that was within a collection has been deleted
+    # and the changes are reverted, the component should appear in the
+    # collection again.
+    for component in get_library_components(library_key, in_collections=True):
```
```suggestion
    components_in_collections = authoring_api.get_components(
        learning_package.id, draft=True, namespace='xblock.v1',
    ).filter(publishable_entity__collections__isnull=False)

    for component in components_in_collections:
```
I don't think `in_collections` is a useful feature to expose as the public API, so I suggest refactoring like this to avoid it.

ChrisChV replied:
Updated: https://github.com/openedx/edx-platform/pull/35734/commits/5e771cebf28e066e462fa39e6dac79d713de79f4

pomegranited commented on openedx/core/djangoapps/content/search/documents.py:
```diff
@@ -485,6 +486,19 @@ def searchable_doc_for_collection(
     if collection:
         assert collection.key == collection_key
 
+        draf_num_children = 0
```
nit:
```suggestion
        draft_num_children = 0
```

ChrisChV replied:
Updated: https://github.com/openedx/edx-platform/pull/35734/commits/5e771cebf28e066e462fa39e6dac79d713de79f4

pomegranited commented on openedx/core/djangoapps/content_libraries/tests/test_api.py:
```diff
@@ -561,3 +561,155 @@ def test_set_library_component_collections(self):
+            event_receiver.call_args_list[1].kwargs,
+        )
+
+    def test_delete_component_and_revert(self):
```
Thank you for testing this case too @ChrisChV !

pomegranited reviewed (approved):
:+1: Works perfectly, thank you for dealing with all these edge cases @ChrisChV !

- [x] I tested this using (very complete) the PR test instructions
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] <strike>Includes documentation</strike> N/A
- [x] <strike>User-facing strings are extracted for translation</strike> N/A

ormsbee commented on requirements/edx/development.txt:
```diff
@@ -1381,7 +1381,7 @@ openedx-filters==1.11.0
     #   lti-consumer-xblock
     #   ora2
-openedx-learning==0.17.0
+openedx-learning @ git+https://github.com/open-craft/openedx-learning.git@chris/FAL-3921-get-drafts-and-get-published
```
Now that openedx-learning has been updated in PyPI, please remove the direct git references here.

ChrisChV replied:
Done, updated here: https://github.com/openedx/edx-platform/pull/35734/commits/f9ef3f9649e58a45a53324f9211fd4e888fb9d82

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1318,6 +1341,39 @@ def revert_changes(library_key):
+    #
+    # This is to update component counts in all library collections,
+    # because there may be components that have been discarded in the revert.
+    for collection in authoring_api.get_collections(learning_package.id):
```
I'm not going to block this PR over this, but I don't think re-indexing every Collection in the Library is an acceptable long term solution. A Library might have many, many Collections. One of the deep problems with Modulestore has been the tendency for apps to have to reprocess the entire course because they don't have granular data on whether the pieces that are relevant to them have changed or not. I don't want that pattern to get started here.

The `reset_drafts_to_published` function can be modified to return a queryset of `Drafts` or `PublishableEntities` or something along those lines, and we can allow this code to figure out which collections were actually updated. (Side note: The `reset_drafts_to_published` is also pretty inefficient because I was only thinking about the manual editing use case and not really thinking about importing potentially thousands of components at once and then hitting "discard changes".)

ChrisChV replied:
I agree with this, could this be done in a follow-up task? @bradenmacdonald @pomegranited

bradenmacdonald replied:
Yes, as long as the bug is fixed now and the fix is backported to Sumac, we can address those refactors and improvements in a follow-up task. Let's do that as soon as sumac.0 is cut. Doing it now will complicate other backports like this one, especially if any change to openedx-learning are required.

But I don't want to postpone it any later than that, or we won't get to it. Ideally you can start on the PRs for it now, even if we don't merge until after Sumac.0 is released.

ormsbee commented on openedx/core/djangoapps/content_libraries/api.py:
```diff
@@ -1318,6 +1341,39 @@ def revert_changes(library_key):
         )
     )
 
+    # For each collection, trigger LIBRARY_COLLECTION_UPDATED signal and set background=True to trigger
```
[off-topic] Out of scope, but also wanted to note that `revert_changes` might not be the best name for this function. It's described as "Discard Changes" in the UI, and it means specifically to reset the draft version to be the published version for all content. The word "revert" might be more appropriate to use when you have something that was already published, and you want to undo that publish. That's definitely a more product question though.

ormsbee reviewed (changes requested):
Some longer term observations, but the only blocking thing is to please change the pinned dependency references to not use the GitHub repo urls.

ormsbee commented on requirements/common_constraints.txt:
```diff
@@ -26,13 +25,6 @@ elasticsearch<7.14.0
-# event-tracking 2.4.1 upgrades to pymongo 4.4.0 which is not supported on edx-platform.
-# We will pin event-tracking to do not break existing installations
-# This can be unpinned once https://github.com/openedx/edx-platform/issues/34586
-# has been resolved and edx-platform is running with pymongo>=4.4.0
```
Was this pulled in accidentally?

ChrisChV replied:
This was introduced in https://github.com/openedx/edx-platform/pull/35734/commits/f9ef3f9649e58a45a53324f9211fd4e888fb9d82. I ran `make compile-requirements` in the cms shell. The command generates those changes automatically. Is this something unexpected? I have re-run the command on this branch and `master` and I get the same result

bradenmacdonald replied:
Personally, I always remove any changes like that from the diff before committing, if it's unrelated to the PR in question. Keep things focused on one set of changes. Those dependencies will be updated in a separate PR.

ChrisChV replied:
Ok, makes sense, I will do that :+1:

ChrisChV replied:
@bradenmacdonald @ormsbee It's already updated here https://github.com/openedx/edx-platform/commit/040ff12430d2418f8bea0c7ec8d2eeb0f734a8e7

ChrisChV commented:
Hi @ormsbee, this is ready for another review

ormsbee reviewed (approved):
@ChrisChV: Please squash your commits and add context from the PR description into the commit message. I'll merge this in the morning EST.

ChrisChV committed:
fix: Keep sync component count in collection card

Fixed component counter synchronization in these cases:

* When deleting a component inside a collection.
* With the library published, when adding a new component in a collection and reverting library changes.
* With the library published, when deleting a component inside a collection and reverting library changes.

Also adds a published > num_counts field in collections in the search index.

ChrisChV commented:
@ormsbee Done :+1:
